File: javascript/initDatabase.js
const fs = require("fs");
const sqlite3 = require("sqlite3").verbose();
const { User, Post, Comment } = require("./classes.js");

let db = new sqlite3.Database("./data/database.db", sqlite3.OPEN_READWRITE);

function createTables() {
	db.run(`
        CREATE TABLE IF NOT EXISTS users(
            userID INT PRIMARY KEY, 
            name TEXT)
        `);

	db.run(`
        CREATE TABLE IF NOT EXISTS posts(
            postID INT PRIMARY KEY,
            title TEXT,
            body TEXT,
            userID INT,
            FOREIGN KEY (userID) references users(userID))
        `);

	db.run(`
        CREATE TABLE IF NOT EXISTS comments(
            commentID INT PRIMARY KEY,
            text TEXT,
            userID INT,
            postID INT,
            FOREIGN KEY (userID) references users(userID),
            FOREIGN KEY (postID) references posts(postID))
        `);
}

//createTables()

function writeToDatabase() {
	const rawdata = fs.readFileSync("./data/sampleData.json");
	const jsonData = JSON.parse(rawdata);

	const userRecords = jsonData.users;
	const postRecords = jsonData.posts;
	const commentRecords = jsonData.comments;

	userRecords.forEach((record) => {
		const insert = `INSERT INTO users (userID, name) values (?, ?)`;
		db.run(insert, [record.UserID, record.Name], (err) => {
			if (err)
				console.log(`couldn't log user: ${record.UserID} into the database`);
		});
	});

	postRecords.forEach((record) => {
		const insert = `INSERT INTO posts (postID, title, body, userID) values (?, ?, ?, ?)`;
		db.run(
			insert,
			[record.PostID, record.Title, record.Body, record.UserID],
			(err) => {
				if (err)
					console.log(`couldn't log post: ${record.PostID} into the database`);
			}
		);
	});

	commentRecords.forEach((record) => {
		const insert = `INSERT INTO comments (commentID, text, postID, userID) values (?, ?, ?, ?)`;
		db.run(
			insert,
			[record.CommentID, record.Text, record.PostID, record.UserID],
			(err) => {
				//prettier-ignore
				if (err) console.log(`couldn't log comment: ${record.CommentID} into the database`);
			}
		);
	});
}

//writeToDatabase()

File: python/initDatabase.py
import sqlite3
import json

conn = sqlite3.connect('./data/database.db')
cursor = conn.cursor()

def init_tables():
    cursor.execute('''
                CREATE TABLE IF NOT EXISTS users
                    (
                    userid INT PRIMARY KEY,
                    name   TEXT
                    ) 
                ''')
    cursor.execute('''
                CREATE TABLE IF NOT EXISTS posts
                    (
                    postID INT PRIMARY KEY,
                    title TEXT,
                    body TEXT,
                    userID INT,
                    FOREIGN KEY (userID) references users(userID)
                    )
                ''')
    cursor.execute('''
               CREATE TABLE IF NOT EXISTS comments
                (
                commentID INT PRIMARY KEY,
                text TEXT,
                userID INT,
                postID INT,
                FOREIGN KEY (userID) references users(userID),
                FOREIGN KEY (postID) references posts(postID)
                ) 
               ''')

def write_to_database():
    with open('./data/sampleData.json') as file:
        data = json.load(file)

    users = []
    posts = []
    comments = []

    for key, records in data.items():
        for record in records:
            if(key == "users"):
                users.append(record)
            elif(key == "posts"):
                posts.append(record)
            else:
                comments.append(record)

    for user in users:
        cursor.execute('''INSERT INTO users (userID, name) values (?, ?)''', (user["UserID"], user["Name"]))

    for post in posts:
        cursor.execute('''INSERT INTO posts (postID, title, body, userID) values (?, ?, ?, ?)''', 
                    (post["PostID"], post["Title"], post["Body"], post["UserID"]))
        
    for comment in comments:
        cursor.execute('''INSERT INTO comments (commentID, text, postID, userID) values (?, ?, ?, ?)''', 
                    (comment["CommentID"], comment["Text"], comment["PostID"], comment["UserID"]))

    conn.commit()

def main():
    print("nothing to do here :)")
    conn.close()


if __name__ == "__main__":
    main()

